docs: Update Key IDs and generator sections to reference valid AWS KMS values, additional naming clarity (#123)
awslabs/aws-encryption-sdk-specification: framework/kms-keyring.md

## Version

> 0.3.0

Author: Yup! I'll have conflicts between the two PR's anyways, so I was just going to update the number during the conflict resolution.

### Changelog

> - 0.3.0

Reviewer: This looks like a Z version bump to me.

Author: Updating.

>   - Rename Key IDs to [Key Names](#key-names) for increased clarity

Reviewer 2: I know why you did not link to the issues,
but should we at least link to the PR?
I know we can get it from source, but it is nice
and easy to have it handy right in the doc?

If the client supplier cannot supply a client for the requested region,
it MUST communicate that fact to the KMS keyring.

### Key Names

> The key name identifies the [AWS KMS customer master keys (CMKs)](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id)

Reviewer: [mixed plurality]
"key name" vs "identifies the CMKs that the keyring uses"

As the spec is currently defined, this is an input that accepts multiple values, each value being a key name.

Author: Good catch. Updating to "Key names identify..."

that the AWS KMS keyring uses for data key encryption and decryption.

> The AWS KMS keyring accepts all [key identifiers accepted by AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id) as key names.

Reviewer: The actor here feels a bit muddled. Maybe:
```suggestion
The AWS KMS keyring accepts any value for a key name that is [accepted by AWS KMS as  a key identifer](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id).
```

There's also still an implied statement here that the keyring *does not* accept values that *are not* accepted by AWS KMS, which is not something that we can guarantee...I'm not sure what the right answer is for that.

Author: I like the change. Updating to it locally.

I think it might make sense to add a statement along the lines of "Key names are passed as-is to AWS KMS, without additional validation" or something else to be very clear.

> Note that only key names in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) will ever be used for decryption.

Reviewer: ```suggestion
Note that only key names in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) are used for decryption.
```

Author: Updating.

> This is because encrypted data keys constructed by the AWS KMS keyring will always store the ID of the

Reviewer: ```suggestion
This is because encrypted data keys constructed by the AWS KMS keyring always store the ID of the
```

Author: Updating

> CMK used to encrypt it in key ARN format, and [OnDecrypt](#ondecrypt) checks the key name against that

Reviewer: ```suggestion
CMK used to encrypt it in key ARN format and [OnDecrypt](#ondecrypt) checks the key name against that
```

Author: I think the comma here is actually correct since the two parts are each independent clauses. But the first sentence is a bit weird. Going to fix this.

> value before attempting decryption.

Reviewer: I feel like we could simplify this to state that the value that AWS KMS keyring writes as the provider info in the serialized EDK is the value that it receives in the response from AWS KMS, then mention as an aside that this value happens to be the CMK ARN.

Author: I wasn't really sure how to properly phrase this. I update the sentences we had here to make them(hopefully) clearer, but wording recommendations here are appreciated.

> The AWS KMS CMK specified by the key names MUST have

Reviewer: [mixed plurality]
"key names" -> plural
"AWS KMS CMK" -> singular
```suggestion
Each AWS KMS CMK specified in the key names MUST have
```

Author: Updating.

[kms:Encrypt](https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html#AWS-KMS-API-Operations-and-Permissions)

> permissions.

Reviewer: Out of scope for this PR, but this wording confuses the actor and subject of the permissions.

I created #125 to track this.

Author: Good note!

Author: I can update the usage here as "The AWS KMS caller using the ..." or "The AWS KMS caller using each..."

### Generator

> A string that identifies an [AWS KMS customer master key (CMK)](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id) that the AWS KMS keyring uses for data key generation,
> as well as data key encryption and decryption.

Reviewer: ```suggestion
A string that identifies an [AWS KMS customer master key (CMK)](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id)
that the AWS KMS keyring uses to generate, encrypt, and decrypt data keys.
```

Author: Updating here and in Key Names (key names without "generate").

> The AWS KMS keyring accepts any [key identifier accepted by AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id) as a generator.

Reviewer: ditto on the confused actor (see "key names" comments)

Author: Already updated :)

> Note that only generators in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) will ever be used for decryption.

Reviewer: ```suggestion
Note that only generators in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) are used for decryption.
```

Author: Updated.

This is because encrypted data keys constructed by the AWS KMS keyring will always store the ID of the

> CMK used to encrypt it in key ARN format, and [OnDecrypt](#ondecrypt) checks the generator against that

Reviewer: See comments on this bit in "key names"

Author: Already updated these sentences.

If the client supplier cannot supply a client for the requested region,
it MUST communicate that fact to the AWS KMS keyring.

### Key Names

A list of strings identifying the [AWS KMS customer master keys (CMKs)](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id)
that the AWS KMS keyring uses to encrypt and decrypt data keys.

The AWS KMS keyring accepts any value for a key name that is [accepted by AWS KMS as a key identifer](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id).
The AWS KMS keyring does not validate any value for a key name and passes this information to AWS KMS.

Note that only key names in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) are used for decryption

> because encrypted data keys constructed by the AWS KMS keyring always store the ID of the CMK used to encrypt it in the key ARN format.
> [OnDecrypt](#ondecrypt) checks the key name against that

Reviewer: This reads awkwardly when broken up like this.

```suggestion
because encrypted data keys constructed by the AWS KMS keyring always store the ID of the CMK used to encrypt it in the key ARN format
and [OnDecrypt](#ondecrypt) checks the key name against that
```

value before attempting decryption.

> The AWS KMS caller using each AWS KMS CMK specified in the key names
> MUST have [kms:Encrypt](https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html#AWS-KMS-API-Operations-and-Permissions) permissions.

Reviewer: ```suggestion
The AWS KMS caller
MUST have [kms:Encrypt](https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html#AWS-KMS-API-Operations-and-Permissions) permission
for each AWS KMS CMK specified in the key names.
```

### Generator

A string that identifies an [AWS KMS customer master key (CMK)](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id)
that the AWS KMS keyring uses to generate, encrypt, and decrypt data keys.

The AWS KMS keyring accepts any value for a generator that is [accepted by AWS KMS as a key identifer](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#key-id).
The AWS KMS keyring does not validate any value for a generator and passes this information to AWS KMS.

Note that only generators in the [key ARN format](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-kms) are used for decryption

> because encrypted data keys constructed by the AWS KMS keyring always store the ID of the CMK used to encrypt it in the key ARN format.
> [OnDecrypt](#ondecrypt) checks the generator against that

Reviewer: ```suggestion
because encrypted data keys constructed by the AWS KMS keyring always store the ID of the CMK used to encrypt it in the key ARN format
and [OnDecrypt](#ondecrypt) checks the generator against that
```

value before attempting decryption.

The generator SHOULD NOT be included in the [key names](#key-names),
otherwise this CMK is used to encrypt the plaintext data key twice.

> The AWS KMS caller using the AWS KMS CMK specified by the generator
> MUST have [kms:GenerateDataKey](https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html#AWS-KMS-API-Operations-and-Permissions) permissions.

Reviewer: ```suggestion
The AWS KMS caller
MUST have [kms:GenerateDataKey](https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html#AWS-KMS-API-Operations-and-Permissions) permission
for the AWS KMS CMK specified by the generator.
```

### Is Discovery

Indicates whether this keyring is a discovery keyring.

> Discovery keyrings do not perform encryption and on decryption attempt to decrypt every inputted
> [encrypted data key](structures.md#encrypted-data-key) if the client supplier return a client.

Reviewer: ```suggestion
Discovery keyrings do not perform encryption and on decryption attempt to decrypt
any [encrypted data key](structures.md#encrypted-data-key) that was encrypted by the AWS KMS keyring,
if the client supplier returns a client.
```
